components/raftstore-v2/src/raft/peer.rs

```diff
@@ -849,4 +858,50 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
             Err(e) => slog_panic!(self.logger, "failed to load term"; "index" => idx, "err" => ?e),
         }
     }
+
+    #[inline]
+    pub fn abnormal_peer_context_mut(&mut self) -> &mut AbnormalPeerContext {
+        &mut self.abnormal_peer_context
+    }
+
+    #[inline]
+    pub fn abnormal_peer_context(&self) -> &AbnormalPeerContext {
+        &self.abnormal_peer_context
+    }
+
+    pub fn any_new_peer_catch_up(&mut self, peer_id: u64) -> bool {
+        // no pending or down peers
+        if self.abnormal_peer_context.is_empty() {
+            return false;
+        }
+        if !self.is_leader() {
+            self.abnormal_peer_context.reset();
+            return false;
+        }
+
+        if self.abnormal_peer_context.is_peer_down(peer_id) {
```

Comment:
I don't get it. Why return directly?

Reply:
It seems that have the same logic in v1. https://github.com/tikv/tikv/blob/315d402780bf1ba8155aafaf57e1608809cedf1b/components/raftstore/src/store/peer.rs#L2198-L2200


I think when receive a AppendResponse from down_peer. which also means that peer is not down any more. so bring it back and make a heartbeat to pd.

And in heartbeat it will re-trigger `collect_down_peers`.

Reply:
That's not true. down peers will update in this pd heartbeat. because in `ready/mod.rs:276` 
https://github.com/tikv/tikv/blob/7ab1702fd84e500a428aa3d02baae24aa3ba46c4/components/raftstore-v2/src/operation/ready/mod.rs#L274-L276
it will insert a new peer heartbeat. and in `collect_down_peers` to makes sure the from peer won't be collected when it is newly updated.

Reply:
But the calling order is `add_peer_heartbeat` -> `any_new_peer_catch_up` -> `collect_down_peers`, right?

Reply:
yes. is that any problem here?🤔

Reply:
so, maybe down peer didn't match the truncated index too. it should also check the matched index too